src/sql-parser/src/parser.rs

```diff
@@ -3792,16 +3797,27 @@ impl<'a> Parser<'a> {
         Ok(Assignment { id, value })
     }
 
-    fn parse_optional_args(&mut self) -> Result<FunctionArgs<Raw>, ParserError> {
+    fn parse_optional_args(
+        &mut self,
+        allow_order_by: bool,
+    ) -> Result<(FunctionArgs<Raw>, Vec<OrderByExpr<Raw>>), ParserError> {
         if self.consume_token(&Token::Star) {
             self.expect_token(&Token::RParen)?;
-            Ok(FunctionArgs::Star)
+            Ok((FunctionArgs::Star, vec![]))
         } else if self.consume_token(&Token::RParen) {
-            Ok(FunctionArgs::Args(vec![]))
+            Ok((FunctionArgs::Args(vec![]), vec![]))
         } else {
             let args = self.parse_comma_separated(Parser::parse_expr)?;
+            // ORDER BY can only appear after at least one argument, and not after a
+            // star. We can ignore checking for it in the other branches. See:
+            // https://www.postgresql.org/docs/current/sql-expressions.html#SYNTAX-AGGREGATES
+            let order_by = if allow_order_by && self.parse_keywords(&[ORDER, BY]) {
+                self.parse_comma_separated(Parser::parse_order_by_expr)?
+            } else {
+                vec![]
+            };
```

Comment:
FWIW, the `ORDER BY` clause seems syntactically valid for table functions too:

```
benesch=# select 1 from unnest(array[1] order by 2);
ERROR:  ORDER BY specified, but unnest is not an aggregate function
```

I think that is an argument for changing `FunctionArgs` to look like so:

```rust
enum FunctionArgs {
    Star,
    Args {
        args: Vec<Expr>,
        order_by: Vec<OrderByExpr<Raw>>,
    }
}
```

I think that has the nice dual benefit of making the call sites cleaner and making the AST reflect the syntax a bit more precisely.

Reply:
No thoughts about the syntax, but what does `select 1 from unnest(array[1] order by 2)` do? We are going to drop the ordering immediately, right? And, probably SQL spec does to, or is that not the case?

Reply:
Right, it errors immediately with that error above (`ERROR:  ORDER BY specified, but unnest is not an aggregate function`). It's a syntactically valid construction but not a semantically valid one.

Reply:
Done.